README.md
# bifrost

> Be warned, I shall uphold my sacred oath to protect this realm as its
> gatekeeper. If your return threatens the safety of Asgard, my gate will
> remain shut and you will be left to perish on the cold waste of Jotunheim.
>   \- Heimdall

A Clojure library for building HTTP API gateways using Pedestal and core.async.

[![Build Status](https://travis-ci.org/democracyworks/bifrost.svg?branch=master)](https://travis-ci.org/democracyworks/bifrost)

## Usage

Add the library as a dependency in your project.clj:

```clojure
[democracyworks/bifrost "0.3.0"]
```

Require the core namespace in your pedestal service namespace:

```clojure
(ns my-gateway.service
  (:require [bifrost.core :as bifrost]
            [clojure.core.async :as async))
```

Then you can use the `bifrost/interceptor` function to create
core.async-backed interceptors. You can pass either function that
returns a core.async channel, or a channel that you are handling
otherwise.

For example:

```clojure
(defn lookup-from-api [bifrost-request]
  (let [return-channel (async/promise-chan)]
    (async/thread
      (async/put! return-channel {:status :ok :lookup (long-running-request)))
    return-channel))

(def lookup-interceptor (bifrost/interceptor lookup-from-api))
```

Or:

```clojure
(def fibonacci-service (async/chan))

(async/go
  (loop [[return-channel bifrost-request] (async/<! fibonacci-service)]
    (->> bifrost-request
         :number
         calculate-fib
         (assoc {:status :ok} :fib)
         (async/put! return-channel))))

(def fibonnaci-interceptor (bifrost/interceptor fibonacci-service))
```

Bifrost interceptors have `enter` and `leave` functions.

When created with a channel-returning function, `enter` will call that
function with the bifrost request.

When created with a async channel, `enter` will put a message on that
channel. The message is a two-element vector. The first element is a
channel to put a response onto, and the second element is the
context's request transformed into a bifrost request.

Either way, the enter function returns the context with the response
channel as a key on `:response-channels`.

The `leave` function attempts to take from the response channel. If
it's already closed (e.g., if you've provided another interceptor in
the chain that deals with its contents) it will forward on the context
without change. If it can take from the response channel, it will
transform the response into a context with a response and merge it
with the incoming context. If it cannot take from the response channel
within the timeout, it will create a 504 response.

For interceptors that should have exclusive access to the response channel, use
`handler` instead, which assocs a 500 response when the response channel closes
unexpectedly.

Bifrost requests are Pedestal requests with the following changes:

* On GET and DELETE
  * Merges `:bifrost-params` -> `:path-params` -> `:query-params` (so
    `:bifrost-params` clobbers `:path-params` which clobbers
    `:query-params`) and puts the resulting map into a vector like:
    `[response-channel params-map]`. (The `response-channel` is
    created for you.)
* On POST, PUT, and PATCH
  * Merges `:bifrost-params` -> `:path-params` -> `:body-params` ->
    `:json-params` -> `:transit-params` -> `:edn-params` ->
    `:form-params` -> `:query-params` and puts the resulting map into
    a vector like: `[response-channel params-map]`. (The
    `response-channel` is created for you.)

Responses should be EDN maps that look like the following:

```clojure
{:status :ok} ; plus any other keys and values you'd like to add

;; or for errors
{:status :error
 :error {:type :timeout}}
```

`:status` can be any of:

* `:ok` - HTTP 200 response
* `:created` - HTTP 201 response
* `:error` - based on the `:type` key in the `:error` map

`:error` statuses may optionally include an `:error` map with a `:type` key.
In this case, the HTTP response is determined by `:type`

* `:semantic` - HTTP 400 response
* `:validation` - HTTP 400 response
* `:not-found` - HTTP 404 response
* `:server` - HTTP 500 response
* `:timeout` - HTTP 504 response
* without a `:type` - HTTP 500 response
* (more to come)

If you want to add anything else to the `params-map`, then just put an
interceptor in front of your core.async interceptor that adds keys & values to
the `[:request :bifrost-params]` key path in the context.
Bifrost will then merge the `:bifrost-params` value into everything
else so those keys will clobber the others in the final `params-map`.

### Update Interceptors

The `bifrost.interceptors` namespace includes some handy interceptors for
transforming your requests and responses before and after bifrost relays
them over the core.async channels in your routes.

Specifically:

* `update-in-request`
    * 2-arity version takes a key path (like `get-in`) and a function.
    * 3+-arity version takes a source key path, a target key path, a function,
    and any additional args to that function.
    * It returns a before interceptor that gets the value at (source-)key-path
    in the request, calls the function argument on it as the first arg, and then
    assoc's the return value into the request at (target-)key-path.
    * If the target-key-path is different than the source-key-path, the last key
    in source-key-path will be dissoc'd from the request data structure.
* `update-in-response`
    * Just like `update-in-request` but returns an after interceptor that
    operates on the response instead of the request.

Use of these interceptors is optional.

## License

Copyright © 2015-2017 Democracy Works, Inc.

Distributed under the Mozilla Public License either version 2.0 or (at
your option) any later version.


ChangeLog.md
# Change Log

## Changes between 0.2.0 and 0.3.0

* **[Breaking]**: removed the lower-level `bifrost.core/async-interceptor`
  function.
* Added a new `bifrost.core/handler` function that works like
  `bifrost.core/interceptor`, but assocs an error response if the channel is
  already closed.

## Changes between 0.1.5 and 0.2.0

* `bifrost.core/interceptor` is no longer a macro.
* The first argument `bifrost.core/interceptor`, instead of being an
  async channel, may be a function that will called with the bifrost
  request and returns a core async channel that the result will
  eventually be placed on.

## Changes between 0.1.4 and 0.1.5

* Added a new optional argument to `bifrost.core/interceptor` and
`bifrost.core/async-interceptor` called `timeout` that allows specifying how
long bifrost should wait for a RabbitMQ response before giving up and putting a
504 Gateway Timeout response into the context. Timeouts are specified in
milliseconds. If not provided, they default to 10,000 milliseconds, just like
before. This should be a non-breaking change.

## Changes between 0.1.3 and 0.1.4

* Fixed a bug in the `bifrost.interceptors/update-*` interceptors where they
assumed that the updated value would always be a map and thus could have `dissoc`
called on it. But that's a bad assumption and bifrost now checks whether or not
it's a map instead of crashing with a cryptic cast exception.

## Changes between 0.1.2 and 0.1.3

* Updated a test for the 0.1.2 change timeout status change.

## Changes between 0.1.1 and 0.1.2

* In 0.1.1 timeouts returned 500 statuses, but 504 is probably the more
appropriate status to return. So 0.1.2 does that.

## Changes between 0.1.0 and 0.1.1

* 0.1.0 put the entire pedestal request onto the core.async channel, but in
0.1.1 the original intent of just putting the merged params map onto the
channel has been restored.


test/bifrost/core_test.clj
(ns bifrost.core-test
  (:require [clojure.test :refer :all]
            [bifrost.core :refer :all]
            [clojure.core.async :as async]))

(deftest interceptor-with-async-chan-test
  (testing "adds a key to the context where the response channel is"
    (let [key-test-ch (async/chan)
          async-interceptor (interceptor key-test-ch)
          enter (:enter async-interceptor)
          out-ctx (enter {:request {:request-method :get}})]
      (is (= 1 (count (:response-channels out-ctx))))))
  (testing "puts just the bifrosted request onto the request part of the channel"
    (let [request-only-test-ch (async/chan)
          async-interceptor (interceptor request-only-test-ch)
          enter (:enter async-interceptor)
          request {:request-method :get
                   :query-params {:test true}}
          ctx {:request request}]
      (enter ctx)
      (let [[response-ch request] (async/<!! request-only-test-ch)]
        (is (= request
               (ctx->bifrost-request ctx))))))
  (testing "takes a API-like response from the response-ch and puts a Ring-like response on the ctx"
    (let [response-test-ch (async/chan)
          async-interceptor (interceptor response-test-ch)
          {:keys [enter leave]} async-interceptor
          request {:request-method :get
                   :bifrost-params {:test true}}
          ctx {:request request}
          middle-ctx (enter ctx)]
      (let [[response-ch request] (async/<!! response-test-ch)]
        (async/>!! response-ch {:status :ok :bridge-endpoints #{"Midgard" "Asgard"}})
        (let [final-ctx (leave middle-ctx)]
          (is (= 200 (get-in final-ctx [:response :status])))
          (is (= {:bridge-endpoints #{"Midgard" "Asgard"}} (get-in final-ctx [:response :body])))))))
  (testing "will respond with a timeout if nothing happens on the response ch"
    (let [timeout-test-ch (async/chan)
          async-interceptor (interceptor timeout-test-ch)
          {:keys [enter leave]} async-interceptor
          request {:request-method :get
                   :bifrost-params {:test true}}
          ctx {:request request}
          middle-ctx (enter ctx)]
      (let [[response-ch request] (async/<!! timeout-test-ch)]
        (let [final-ctx (leave middle-ctx)]
          (is (= 504 (get-in final-ctx [:response :status])))
          (is (= "Bifrost timeout" (get-in final-ctx [:response :body])))))))
  (testing "will forward the ctx through if the response channel has been closed"
    (let [closed-test-ch (async/chan)
          async-interceptor (interceptor closed-test-ch)
          {:keys [enter leave]} async-interceptor
          request {:request-method :get
                   :bifrost-params {:test true}}
          ctx {:request request}
          middle-ctx (enter ctx)
          otherwise-created-ctx (merge middle-ctx
                                       {:response {:status :201 :body "Handled by someone else"}})]
      (let [[response-ch request] (async/<!! closed-test-ch)]
        (async/close! response-ch)
        (let [final-ctx (leave otherwise-created-ctx)]
          (is (= final-ctx otherwise-created-ctx)))))))

(deftest interceptor-with-fn-test
  (let [fn-with-response (fn [_]
                           (async/go
                             {:status :ok
                              :bridge-endpoints #{"Midgard" "Asgard"}}))
        async-identity (fn [arg]
                         (async/go arg))
        fn-timeout (fn [_]
                     (async/chan))
        fn-closed-chan (fn [_]
                         (let [c (async/chan)]
                           ;; fake a situation where the function
                           ;; closes the channel if it knows there's
                           ;; nothing worth doing
                           (async/close! c)
                           c))]
    (testing "adds a key to the context where the response channel is"
      (let [fn-interceptor (interceptor fn-with-response)
            enter (:enter fn-interceptor)
            out-ctx (enter {:request {:request-method :get}})]
        (is (= 1 (count (:response-channels out-ctx))))))
    (testing "calls the function with just the bifrosted request"
      (let [async-interceptor (interceptor async-identity)
            enter (:enter async-interceptor)
            request {:request-method :get
                     :query-params {:test true}}
            ctx {:request request}
            out-ctx (enter ctx)]
        (let [chan (-> out-ctx :response-channels vals first)
              bifrost-request (async/<!! chan)]
          (is (= bifrost-request
                 (ctx->bifrost-request ctx))))))
    (testing "takes a API-like response from the channel the function returns and puts a Ring-like response on the ctx"
      (let [{:keys [enter leave]} (interceptor fn-with-response)
            request {:request-method :get
                     :bifrost-params {:test true}}
            ctx {:request request}
            out-ctx (leave (enter ctx))]
        (is (= 200 (get-in out-ctx [:response :status])))
        (is (= {:bridge-endpoints #{"Midgard" "Asgard"}}
               (get-in out-ctx [:response :body])))))
    (testing "will respond with a timeout if nothing happens on the channel returned by the function"
      (let [{:keys [enter leave]} (interceptor fn-timeout)
            request {:request-method :get
                     :bifrost-params {:test true}}
            ctx {:request request}
            out-ctx (leave (enter ctx))]
        (is (= 504 (get-in out-ctx [:response :status])))
        (is (= "Bifrost timeout" (get-in out-ctx [:response :body])))))
    (testing "will forward the ctx through if the channel returned by the function has been closed"
      (let [{:keys [enter leave]} (interceptor fn-closed-chan)
            request {:request-method :get
                     :bifrost-params {:test true}}
            ctx {:request request}
            middle-ctx (enter ctx)
            otherwise-created-ctx (merge middle-ctx
                                         {:response
                                          {:status :201
                                           :body "Handled by someone else"}})
            out-ctx (leave otherwise-created-ctx)]
        (is (= out-ctx otherwise-created-ctx))))))

(deftest handler-with-fn-test
  (let [fn-with-response (fn [_]
                           (async/go
                             {:status :ok
                              :bridge-endpoints #{"Midgard" "Asgard"}}))
        async-identity (fn [arg]
                         (async/go arg))
        fn-timeout (fn [_]
                     (async/chan))
        fn-closed-chan (fn [_]
                         (let [c (async/chan)]
                           ;; fake a situation where the function
                           ;; closes the channel if it knows there's
                           ;; nothing worth doing
                           (async/close! c)
                           c))]
    (testing "adds a key to the context where the response channel is"
      (let [fn-interceptor (handler fn-with-response)
            enter (:enter fn-interceptor)
            out-ctx (enter {:request {:request-method :get}})]
        (is (= 1 (count (:response-channels out-ctx))))))
    (testing "calls the function with just the bifrosted request"
      (let [async-interceptor (handler async-identity)
            enter (:enter async-interceptor)
            request {:request-method :get
                     :query-params {:test true}}
            ctx {:request request}
            out-ctx (enter ctx)]
        (let [chan (-> out-ctx :response-channels vals first)
              bifrost-request (async/<!! chan)]
          (is (= bifrost-request
                 (ctx->bifrost-request ctx))))))
    (testing "takes a API-like response from the channel the function returns and puts a Ring-like response on the ctx"
      (let [{:keys [enter leave]} (handler fn-with-response)
            request {:request-method :get
                     :bifrost-params {:test true}}
            ctx {:request request}
            out-ctx (leave (enter ctx))]
        (is (= 200 (get-in out-ctx [:response :status])))
        (is (= {:bridge-endpoints #{"Midgard" "Asgard"}}
               (get-in out-ctx [:response :body])))))
    (testing "will respond with a timeout if nothing happens on the channel returned by the function"
      (let [{:keys [enter leave]} (handler fn-timeout)
            request {:request-method :get
                     :bifrost-params {:test true}}
            ctx {:request request}
            out-ctx (leave (enter ctx))]
        (is (= 504 (get-in out-ctx [:response :status])))
        (is (= "Bifrost timeout" (get-in out-ctx [:response :body])))))
    (testing "will respond with a closed channel response if the channel returned by the function has been closed"
      (let [{:keys [enter leave]} (handler fn-closed-chan)
            request {:request-method :get
                     :bifrost-params {:test true}}
            ctx {:request request}
            out-ctx (leave (enter ctx))]
        (is (= 500 (get-in out-ctx [:response :status])))
        (is (= "Channel was closed unexpectedly" (get-in out-ctx [:response :body])))))))

(deftest params-map-test
  (testing "GET/DELETE merges bifrost-params -> path-params -> query-params"
    (are [request-method]
        (= {:one 1, :two 2, :three 3, :b "bifrost", :p "path", :q "query"}
           (params-map {:request-method request-method
                        :bifrost-params {:one 1, :two 2, :three 3, :b "bifrost"}
                        :path-params {:two "two", :one "one", :p "path"}
                        :query-params {:one "uno", :q "query"}}))
      :get :delete))
  (testing "POST/PUT/PATCH merges bifrost-params -> path-params -> body-params ->
            json-params -> transit-params -> edn-params -> form-params ->
            query-params"
    (are [request-method]
        (= {:one 1, :two 2, :three 3, :bi "bifrost", :p "path", :q "query",
            :bo "body", :j "json", :t "transit", :e "edn", :f "form"}
           (params-map {:request-method request-method
                        :bifrost-params {:one 1, :two 2, :three 3, :bi "bifrost"}
                        :path-params {:two "two", :one "one", :p "path"}
                        :body-params {:bo "body"}
                        :json-params {:j "json"}
                        :transit-params {:t "transit"}
                        :edn-params {:e "edn"}
                        :form-params {:f "form"}
                        :query-params {:one "uno", :q "query"}}))
      :post :put :patch)))


src/bifrost/interceptors.clj
(ns bifrost.interceptors
  (:require [io.pedestal.interceptor :refer [interceptor]]))

(defn update-interceptor
  [type source-key-path target-key-path f & args]
  (interceptor
   (let [context-key (case type
                       :enter :request
                       :leave :response)]
     {type
      (fn [ctx]
        (let [source-context-path (cons context-key source-key-path)
              target-context-path (cons context-key target-key-path)
              old-value (get-in ctx source-context-path)
              new-value (apply f old-value args)
              new-ctx (assoc-in ctx target-context-path new-value)
              source-parent-keypath (butlast source-context-path)]
          (if (or (= source-context-path target-context-path)
                  (not (map? (get-in new-ctx source-parent-keypath))))
            new-ctx
            (update-in new-ctx source-parent-keypath
                       dissoc (last source-context-path)))))})))

(defn update-in-request
  ([key-path f] (update-in-request key-path key-path f))
  ([source-key-path target-key-path f & args]
   (apply update-interceptor :enter source-key-path target-key-path f args)))

(defn update-in-response
  ([key-path f] (update-in-response key-path key-path f))
  ([source-key-path target-key-path f & args]
   (apply update-interceptor :leave source-key-path target-key-path f args)))


src/bifrost/core.clj
(ns bifrost.core
  (:require [io.pedestal.interceptor :as interceptor]
            [clojure.core.async :as async]
            [ring.util.response :as ring-resp]))

(def default-timeout 10000)

(defn error-response->http-status
  [error-response]
  (case (:type error-response)
    :semantic 400
    :validation 400
    :not-found 404
    :server 500
    :timeout 504
    500))

(defn response->http-status
  [response]
  (case (:status response)
    :ok 200
    :created 201
    :error (error-response->http-status (:error response))
    500))

(defn params-map
  [request]
  (case (:request-method request)
    (:get :delete) (merge (:query-params request)
                          (:path-params request)
                          (:bifrost-params request))
    (:post :put :patch) (merge (:query-params request)
                               (:form-params request)
                               (:edn-params request)
                               (:transit-params request)
                               (:json-params request)
                               (:body-params request)
                               (:path-params request)
                               (:bifrost-params request))))

(defn ctx->bifrost-request [ctx]
  (params-map (:request ctx)))

(defn api-response->ctx
  [api-response]
  (when api-response
    (let [status (response->http-status api-response)]
      {:response (-> api-response
                     (dissoc :status)
                     ring-resp/response
                     (ring-resp/status status))})))

(def timeout-response
  {:response {:status 504
              :body "Bifrost timeout"}})

(def channel-closed-response
  {:response {:status 500
              :body "Channel was closed unexpectedly"}})

(defn response-channel-key []
  (keyword (gensym "bifrost-response-channel-")))

(defn interceptor*
  "The main interceptor logic. Saves a channel on the context in the `:enter`
  handler, and tries to take off of it in the `:leave` handler. Assocs a 504
  response if `timeout` is reached before the channel has a result.

  `->chan` is a function taking a context and returning an async channel that
  will eventually contain a response.

  `->response` is a function taking a (possibly nil) response that returns data
  to merge into the context."
  [->chan ->response timeout]
  (let [response-channel-key (response-channel-key)]
    (interceptor/interceptor
     {:name ::interceptor
      :enter
      (fn [ctx]
        (assoc-in ctx [:response-channels response-channel-key]
                  (->chan ctx)))
      :leave
      (fn [ctx]
        (let [response (async/alt!!
                        (async/timeout timeout)
                        timeout-response

                        (get-in ctx [:response-channels response-channel-key])
                        ([r]
                         (->response r)))]
          (merge ctx response)))})))

(defn fn-interceptor
  [f timeout]
  (interceptor* (comp f ctx->bifrost-request)
                api-response->ctx
                timeout))

(defn- channel-adapter
  "Takes an async `channel` and returns a function that can be used with
  `interceptor*`."
  [channel]
  (let [request-ch (async/chan 1)]
    (async/pipe request-ch channel)
    (fn [ctx]
      (let [response-channel (async/chan 1)]
        (async/>!! request-ch [response-channel ctx])
        response-channel))))

(defn interceptor
  "Returns an async channel-based interceptor. Takes either a function
  returning a response channel, or a channel that accepts requests of the form
  `[response-channel ctx]` and places a single response on `response-channel`.

  Three results are possible:
  1. A response is available within `timeout`: coerces the response to a ring
     response and assocs it onto the context.
  2. A response is not available within `timeout`: assocs `timeout-response`
     onto the context.
  3. The channel closes without a response: returns context as-is on the
     assumption that another interceptor handled the response."
  ([fn-or-chan]
   (interceptor fn-or-chan default-timeout))
  ([fn-or-chan timeout]
   (if (fn? fn-or-chan)
     (fn-interceptor fn-or-chan timeout)
     (fn-interceptor (channel-adapter fn-or-chan) timeout))))

(defn fn-handler
  [f timeout]
  (interceptor* (comp f ctx->bifrost-request)
                (fn [response]
                  (or (api-response->ctx response)
                      channel-closed-response))
                timeout))

(defn handler
  "Like `interceptor`, but instead of ignoring closed channels, assocs
  `channel-closed-response` onto the context.

  This assumes that no other interceptor has access to the response channel.
  Generally speaking this should be used as a pedestal handler (i.e. last
  interceptor in the chain) to insure that no other interceptor has touched the
  response channel."
  ([fn-or-chan]
   (handler fn-or-chan default-timeout))
  ([fn-or-chan timeout]
   (if (fn? fn-or-chan)
     (fn-handler fn-or-chan timeout)
     (fn-handler (channel-adapter fn-or-chan) timeout))))
